packages/appium/lib/appium.js

```diff
@@ -241,6 +252,330 @@ class AppiumDriver extends DriverCore {
+   * @param {import('@appium/types').ExternalDriver | AppiumDriver} bidiHandlerDriver - the driver
+   * handling the bidi commands
+   * @param {(data: string | Buffer) => Promise<void>} send - a method used to send data to the
+   * client
+   */
+  initBidiEventListeners(ws, bidiHandlerDriver, send) {
+    // If the driver emits a bidi event that should maybe get sent to the client, check to make
+    // sure the client is subscribed and then pass it on
+    let eventListener = async ({context, method, params}) => {
+      // if the driver didn't specify a context, use the empty context
+      if (!context) {
+        context = '';
+      }
+      if (!method || !params) {
+        throw new Error(
+          `Driver emitted a bidi event that was malformed. Require method and params keys ` +
+            `(with optional context). But instead received: ${JSON.stringify({
+              context,
+              method,
+              params,
+            })}`,
+        );
+      }
+      if (ws.readyState !== WebSocket.OPEN) {
+        // if the websocket is not still 'open', then we can ignore sending these events
+        if (ws.readyState > WebSocket.OPEN) {
+          // if the websocket is closed or closing, we can remove this listener as well to avoid
+          // leaks
+          bidiHandlerDriver.eventEmitter.removeListener(BIDI_EVENT_NAME, eventListener);
+        }
+        return;
+      }
+
+      if (bidiHandlerDriver.bidiEventSubs[method]?.includes(context)) {
+        this.log.info(
+          `<-- BIDI EVENT ${method} (context: '${context}', params: ${JSON.stringify(params)})`,
+        );
+        // now we can send the event onto the socket
+        const ev = {type: 'event', context, method, params};
+        // TODO add conditional logic to send only if user is subscribed
+        await send(JSON.stringify(ev));
+      }
+    };
+    bidiHandlerDriver.eventEmitter.on(BIDI_EVENT_NAME, eventListener);
+  }
+
+  /**
+   * @param {Buffer} data
+   * @param {ExternalDriver | AppiumDriver} driver
+   */
+  async onBidiMessage(data, driver) {
+    let resMessage, id, method, params;
+    try {
+      try {
+        ({id, method, params} = JSON.parse(data.toString('utf8')));
+      } catch (err) {
+        throw new errors.InvalidArgumentError('Could not parse Bidi command');
+      }
+      this.log.info(`--> BIDI message #${id}`);
+      if (!method) {
+        throw new errors.InvalidArgumentError('Missing method for BiDi operation');
+      }
+      if (!params) {
+        throw new errors.InvalidArgumentError('Missing params for BiDi operation');
+      }
+      const result = await driver.executeBidiCommand(method, params);
+      // https://w3c.github.io/webdriver-bidi/#protocol-definition
+      resMessage = {
+        id,
+        type: 'success',
+        result,
+      };
+    } catch (err) {
+      resMessage = err.bidiErrObject(id);
+    }
+    this.log.info(`<-- BIDI message #${id}`);
```

Comment:
I believe `this.log` only includes the id of the ubmrella driver, not the id of the actual target driver. If we want to log proper id then we should retrieve the log instance from the actual target driver (e.g. bidiHandlerDriver.log)

Reply:
ah yeah good point